packages/flutter_tools/lib/src/android/java.dart

```diff
@@ -0,0 +1,245 @@
+  // across the  the tool (building Android apps, interacting with the Android SDK, etc.).
+  // Currently, this consistency is fragile since the logic used for building
+  // Android apps exists independently of this method.
+  // See https://github.com/flutter/flutter/issues/124252.
+  static Java? find({
+    required AndroidStudio? androidStudio,
+    required Logger logger,
+    required FileSystem fileSystem,
+    required Platform platform,
+    required ProcessManager processManager,
+  }) {
+    final OperatingSystemUtils os = OperatingSystemUtils(
+      fileSystem: fileSystem,
+      logger: logger,
+      platform: platform,
+      processManager: processManager
+    );
+    final String? home = _findJavaHome(
+      logger: logger,
+      androidStudio: androidStudio,
+      platform: platform
+    );
+    final String? binary = _findJavaBinary(
+      logger: logger,
+      javaHome: home,
+      fileSystem: fileSystem,
+      operatingSystemUtils: os,
+      platform: platform
+    );
+
+    if (binary == null) {
+      return null;
+    }
+
+    return Java(
+      home: home,
+      binary: binary,
+      logger: logger,
+      fileSystem: fileSystem,
+      os: os,
+      platform: platform,
+      processManager: processManager,
+    );
+  }
+
+  /// The path of the runtime's home directory.
+  ///
+  /// This should only be used for logging and validation purposes.
+  /// If you need to set JAVA_HOME when starting a process, consider
+  /// using [getJavaEnvironment] instead.
+  /// If you need to inspect the files of the runtime, considering adding
+  /// a new method to this class instead.
+  final String? home;
+
+  /// The path of the runtime's java binary.
+  ///
+  /// This should be only used for logging and validation purposes.
+  /// If you need to invoke the binary directly, consider adding a new method
+  /// to this class instead.
+  final String binary;
+
+  final Logger _logger;
+  final FileSystem _fileSystem;
+  final OperatingSystemUtils _os;
+  final Platform _platform;
+  final ProcessManager _processManager;
+  final ProcessUtils _processUtils;
+
+  JavaVersion? _version;
+
+  /// Returns an environment variable map with
+  /// 1. JAVA_HOME set if this object has a known home directory, and
+  /// 2. The java binary folder appended onto PATH, if the binary location is known.
+  ///
+  /// This map should be used as the environment when invoking any Java-dependent
+  /// processes, such as Gradle or Android SDK tools (avdmanager, sdkmanager, etc.)
+  Map<String, String> getJavaEnvironment() {
+    return <String, String>{
+      if (home != null) _javaHomeEnvironmentVariable: home!,
+      'PATH': _fileSystem.path.dirname(binary) +
+                        _os.pathVarSeparator +
+                        _platform.environment['PATH']!,
+    };
+  }
+
+  /// Returns the version of java in the format \d(.\d)+(.\d)+
+  /// Returns null if version could not be determined.
+  JavaVersion? getVersion() {
```

Comment:
nit:

```suggestion
  JavaVersion? get version {
```

Reply:
Or even:

```dart
  late final JavaVersion? version = (() {
    final RunResult result = _processUtils.runSync(
      <String>[binary, '--version'],
      environment: getJavaEnvironment(),
    );
    if (result.exitCode != 0) {
      _logger.printTrace('java --version failed: exitCode: ${result.exitCode}'
        ' stdout: ${result.stdout} stderr: ${result.stderr}');
    }
    return JavaVersion.tryParseFromJavaOutput(result.stdout, logger: _logger);
  })();
```

Reply:
and does `java version` run fast enough for us to call it synchronously?

Reply:
I chose a method rather than a property because a method better communicates that some computation may occur, whereas properties subtly imply that the value is pre-computed and that reading it is a simple, fast memory-access. Perhaps I could rename it `findVersion` or `determineVersion` to better communicate this.

Prior to this change, we already ran `java --version` synchronously, so this is not a change.

It runs < 0.04 seconds on my idle Macbook. Also, we only call this during project validation and when running the recently-added `AndroidStudioJavaGradleConflictMigration` migrator, where I don't think performance is highly critical. I also have to wonder if overhead from asynchronism isn't greater than the actual computation time.

Reply:
SGTM, 0.04 seconds is pretty fast :) I was just worried it was like `flutter --version`, which could take 30 seconds :)

Reply:
> I chose a method rather than a property because a method better communicates that some computation may occur

What about the lazily evaluated late final example I provided?

Reply:
I actually didn't realize `late` could be used like this, so that's neat. I think I like this approach since it makes the code more concise. I think the code is fast enough to be a property.

Reply:
So semantically this is equivalent to what you had, it's just cleaner cos you don't need to add a private property and check for it. This is neither faster nor slower.

Reply:
That's true. What I meant is that I still prefer using methods over properties when the computation is non-trivial:

> I chose a method rather than a property because a method better communicates that some computation may occur, whereas properties subtly imply that the value is pre-computed and that reading it is a simple, fast memory-access (or something not much slower than that).